internal/webserver/webserver.go

```diff
@@ -252,35 +252,64 @@ func (n kubeFilter) getOwnedNamespacesForUser(username string) (res []string, er
 	if err := n.client.List(context.Background(), tl, f); err != nil {
 		return nil, fmt.Errorf("cannot retrieve Tenant list: %s", err.Error())
 	}
-
 	for _, t := range tl.Items {
 		res = append(res, t.GetName())
 	}
+	log.V(4).Info("Tenant list", "res", res)
 	return
 }
 
-func (n kubeFilter) getLabelSelectorForUser(username string) (labels.Selector, error) {
+func (n kubeFilter) getTenantsForOwner(ownerKind string, ownerName string) (tenants []string, err error) {
+	tl := &capsulev1alpha1.TenantList{}
+	f := client.MatchingFields{
+		".spec.owner.ownerkind": fmt.Sprintf("%s:%s", ownerKind, ownerName),
+	}
+	if err := n.client.List(context.Background(), tl, f); err != nil {
+		return nil, fmt.Errorf("cannot retrieve Tenants list: %s", err.Error())
+	}
+	for _, t := range tl.Items {
+		tenants = append(tenants, t.GetName())
+	}
+	log.V(4).Info("Tenants list", "tenants", tenants)
+	return
+}
+
+func (n kubeFilter) getLabelSelectorForOwner(username string, groups []string) (labels.Selector, error) {
+	var r *labels.Requirement
 	capsuleLabel, err := capsulev1alpha1.GetTypeLabel(&capsulev1alpha1.Tenant{})
 	if err != nil {
 		return nil, fmt.Errorf("cannot get Capsule Tenant label: %s", err.Error())
 	}
-
-	ownedNamespaces, err := n.getOwnedNamespacesForUser(username)
+	// Find tenants belonging to a user
+	ownedTenants, err := n.getTenantsForOwner("User", username)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get Namespaces slice owned by Tenant Owner: %s", err.Error())
+		return nil, fmt.Errorf("cannot get Tenants slice owned by Tenant Owner: %s", err.Error())
 	}
-
-	var r *labels.Requirement
-
-	if len(ownedNamespaces) > 0 {
-		r, err = labels.NewRequirement(capsuleLabel, selection.In, ownedNamespaces)
-	} else {
-		r, err = labels.NewRequirement("dontexistsignoreme", selection.Exists, []string{})
+	if len(ownedTenants) > 0 {
+		r, err = labels.NewRequirement(capsuleLabel, selection.In, ownedTenants)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse Tenant selector: %s", err.Error())
+		}
+		return labels.NewSelector().Add(*r), nil
```

Comment:
What happens if a Tenant Owner got a Tenant assigned by Name and also a subset of Tenant assigned by the group?

According to this implementation, we would miss the latter ones.

I would suggest refactoring this a little bit, trying to avoid as much as possible the if/else and using a final return statement.

Reply:
Now covered the mentioned case too.